build_resolvers/lib/src/resolver.dart

```diff
@@ -60,11 +61,20 @@ class PerActionResolver implements ReleasableResolver {
       libraries.firstWhere((l) => l.name == libraryName, orElse: () => null);
 
   @override
-  Future<bool> isLibrary(AssetId assetId) => _delegate.isLibrary(assetId);
+  Future<bool> isLibrary(AssetId assetId) =>
+      _resolveIfNecesssary(assetId).then(_delegate.isLibrary);
 
   @override
   Future<LibraryElement> libraryFor(AssetId assetId) =>
-      _delegate.libraryFor(assetId);
+      _resolveIfNecesssary(assetId).then(_delegate.libraryFor);
```

Comment:
We already had a potential for non-hermetic information transfer here. We need to carefully consider whether this makes things any worse.

Reply:
Am I correct to assume that this is the potential problem (with or without the `_resolveIfNecessary` step)?

1. Builder foo is run on `a|lib/foo.dart`, which doesn't import anything
2. Builder bar is run on `a|lib/bar.dart`, but calls `libraryFor(a|lib/foo.dart)`. That works, because the library has already been resolved, but doesn't recognize `foo.dart` as an input.

In that case, keeping a per-buildstep list of assets that have been crawled __and__ wrapping the resolver in `_resolveIfNecessary` would improve the situation, right?

Reply:
Just say the comment that mentions crawling the element model manually. I don't think that gets worse now as the `libraries` stream only crawls from the build step input. Should we include libraries found via subsequent calls to `libraryFor` in that stream?

Reply:
We were talking in person about this exact same question haha. It isn't clear what should happen. It is weird for it to change after you ask for a new library.

I think probably the best situation is if it was a method that took a root AssetId to crawl from, but that would be a breaking change now.

I would probably be fine with leaving it as is and we could add a method like `transitiveLibrariesFor(AssetId)`  or something in the future if people need to crawl libraries from the non-primary input.

Reply:
> The thing that sucks about that though is that `findLibraryByName` won't work for things that aren't imported by the primary input.

What do you think is the best way to solve that? Keep a per-buildstep map of `<String, LibraryElement>` based on crawled assets? We could populate that map in the `BuildAssetUriResolver`.

Reply:
I think including them in the stream if the `libraries` getter has been called _after_ a previous call to `libraryFor` makes sense - after all, the library has been found at that point and is accessible for the resolver. That also means that it will be included in `findLibraryByName`.